File: test-fixtures/sample.ts
import { readFileSync, writeFileSync, existsSync, mkdirSync } from 'node:fs';
import { join, relative, resolve, sep } from 'node:path';
import { createHash, randomBytes } from 'node:crypto';
import { EventEmitter } from 'node:events';

export interface UserConfig {
  name: string;
  email: string;
  role: 'admin' | 'user' | 'viewer';
  settings?: Record<string, unknown>;
}

export interface ApiResponse<T> {
  status: number;
  data: T;
  message: string;
  timestamp: number;
}

export enum HttpMethod {
  GET = 'GET',
  POST = 'POST',
  PUT = 'PUT',
  DELETE = 'DELETE',
  PATCH = 'PATCH',
}

export abstract class BaseService {
  protected baseUrl: string;
  protected timeout: number;

  constructor(baseUrl: string, timeout = 5000) {
    this.baseUrl = baseUrl;
    this.timeout = timeout;
  }

  abstract handle(data: unknown): Promise<ApiResponse<unknown>>;

  protected log(method: string, path: string): void {
    console.log(`[${method}] ${path}`);
  }
}

export class UserService extends BaseService {
  private users: Map<string, UserConfig> = new Map();
  private emitter: EventEmitter = new EventEmitter();

  constructor() {
    super('https://api.example.com/users');
  }

  async handle(data: unknown): Promise<ApiResponse<unknown>> {
    return this.getUser(data as string);
  }

  async getUser(id: string): Promise<ApiResponse<UserConfig>> {
    const user = this.users.get(id);
    if (!user) {
      return { status: 404, data: null as unknown as UserConfig, message: 'Not found', timestamp: Date.now() };
    }
    this.emitter.emit('user:fetched', user);
    return { status: 200, data: user, message: 'OK', timestamp: Date.now() };
  }

  async createUser(config: UserConfig): Promise<ApiResponse<UserConfig>> {
    const id = randomBytes(16).toString('hex');
    this.users.set(id, config);
    this.emitter.emit('user:created', config);
    return { status: 201, data: config, message: 'Created', timestamp: Date.now() };
  }

  async deleteUser(id: string): Promise<ApiResponse<null>> {
    const deleted = this.users.delete(id);
    if (!deleted) {
      return { status: 404, data: null, message: 'Not found', timestamp: Date.now() };
    }
    this.emitter.emit('user:deleted', id);
    return { status: 200, data: null, message: 'Deleted', timestamp: Date.now() };
  }

  on(event: string, handler: (...args: unknown[]) => void): void {
    this.emitter.on(event, handler);
  }
}

function calculateHash(input: string): string {
  return createHash('sha256').update(input, 'utf-8').digest('hex');
}

function validateEmail(email: string): boolean {
  const re = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  return re.test(email);
}

function paginate<T>(items: T[], page: number, perPage: number): { items: T[]; total: number; pages: number } {
  const start = (page - 1) * perPage;
  const end = start + perPage;
  return { items: items.slice(start, end), total: items.length, pages: Math.ceil(items.length / perPage) };
}

function retry<T>(fn: () => Promise<T>, retries: number, delay: number): Promise<T> {
  return fn().catch((err) => {
    if (retries <= 0) throw err;
    return new Promise((resolve) => setTimeout(resolve, delay)).then(() => retry(fn, retries - 1, delay));
  });
}

const DEFAULT_CONFIG: UserConfig = {
  name: 'Guest',
  email: 'guest@example.com',
  role: 'viewer',
};

export { calculateHash, validateEmail, paginate, retry, DEFAULT_CONFIG };

File: test-fixtures/sample.java
package com.example.index;

import java.io.*;
import java.nio.file.*;
import java.security.*;
import java.time.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;
import java.util.stream.*;

public class Sample {

    public enum UserRole {
        ADMIN, USER, VIEWER
    }

    public static class UserConfig {
        private String name;
        private String email;
        private UserRole role;
        private Map<String, Object> settings;

        public UserConfig(String name, String email, UserRole role) {
            this.name = name;
            this.email = email;
            this.role = role;
            this.settings = new HashMap<>();
        }

        public String getName() { return name; }
        public void setName(String name) { this.name = name; }
        public String getEmail() { return email; }
        public void setEmail(String email) { this.email = email; }
        public UserRole getRole() { return role; }
        public void setRole(UserRole role) { this.role = role; }
    }

    public static class ApiResponse<T> {
        private int status;
        private T data;
        private String message;
        private long timestamp;

        public ApiResponse(int status, T data, String message) {
            this.status = status;
            this.data = data;
            this.message = message;
            this.timestamp = Instant.now().toEpochMilli();
        }

        public int getStatus() { return status; }
        public T getData() { return data; }
        public String getMessage() { return message; }
        public long getTimestamp() { return timestamp; }
    }

    public interface Service {
        <T> ApiResponse<T> handle(T data);
    }

    public static class UserService implements Service {
        private final Map<String, UserConfig> users = new ConcurrentHashMap<>();
        private final List<Consumer<UserConfig>> listeners = new CopyOnWriteArrayList<>();

        public ApiResponse<UserConfig> getUser(String id) {
            UserConfig user = users.get(id);
            if (user == null) {
                return new ApiResponse<>(404, null, "Not found");
            }
            listeners.forEach(l -> l.accept(user));
            return new ApiResponse<>(200, user, "OK");
        }

        public ApiResponse<UserConfig> createUser(UserConfig config) {
            if (!validateConfig(config)) {
                return new ApiResponse<>(400, null, "Invalid config");
            }
            String id = generateId();
            users.put(id, config);
            return new ApiResponse<>(201, config, "Created");
        }

        public ApiResponse<UserConfig> updateUser(String id, Map<String, String> updates) {
            UserConfig user = users.get(id);
            if (user == null) {
                return new ApiResponse<>(404, null, "Not found");
            }
            if (updates.containsKey("name")) user.setName(updates.get("name"));
            if (updates.containsKey("email")) user.setEmail(updates.get("email"));
            return new ApiResponse<>(200, user, "Updated");
        }

        public ApiResponse<Void> deleteUser(String id) {
            if (users.remove(id) == null) {
                return new ApiResponse<>(404, null, "Not found");
            }
            return new ApiResponse<>(200, null, "Deleted");
        }

        public List<UserConfig> listUsers() {
            return new ArrayList<>(users.values());
        }

        public void onEvent(Consumer<UserConfig> listener) {
            listeners.add(listener);
        }

        private boolean validateConfig(UserConfig config) {
            if (config.getName() == null || config.getName().isEmpty()) return false;
            if (!config.getEmail().contains("@")) return false;
            return true;
        }

        private String generateId() {
            try {
                MessageDigest md = MessageDigest.getInstance("SHA-256");
                byte[] hash = md.digest(String.valueOf(System.nanoTime()).getBytes());
                StringBuilder hex = new StringBuilder();
                for (byte b : hash) hex.append(String.format("%02x", b));
                return hex.substring(0, 16);
            } catch (NoSuchAlgorithmException e) {
                return UUID.randomUUID().toString();
            }
        }
    }

    public static String calculateHash(String content) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] hash = md.digest(content.getBytes("UTF-8"));
            StringBuilder hex = new StringBuilder();
            for (byte b : hash) hex.append(String.format("%02x", b));
            return hex.toString();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public static boolean validateEmail(String email) {
        return email != null && email.contains("@") && email.contains(".");
    }

    public static <T> PaginatedResult<T> paginate(List<T> items, int page, int perPage) {
        int start = (page - 1) * perPage;
        int end = Math.min(start + perPage, items.size());
        int pages = (int) Math.ceil((double) items.size() / perPage);
        return new PaginatedResult<>(items.subList(start, end), items.size(), pages);
    }

    public static class PaginatedResult<T> {
        public final List<T> items;
        public final int total;
        public final int pages;

        public PaginatedResult(List<T> items, int total, int pages) {
            this.items = items; this.total = total; this.pages = pages;
        }
    }

    public static String readFile(String path) throws IOException {
        return Files.readString(Path.of(path));
    }

    public static void writeFile(String path, String content) throws IOException {
        Files.writeString(Path.of(path), content);
    }
}
